### DISCUSSION TITLE
Performance issues in 3.x versions

### DISCUSSION THREAD

**Discussion by t8m** [openssl/openssl#22900] [General Discussion]
It is a known issue that some application workloads when ported from 1.1.1 version to 3.x versions show significant performance degradation. These are the main reasons for the performance degradation:

1. Added complexity of the library context and accessing the per-libctx data (mostly provider related)
    - this should be mostly resolved in 3.1 version by removals of the ex-data indirection for most of the library context data

2. Added complexity of fetching the provider algorithms
    - application workarounds are possible in most cases by pre-fetching the EVP_MD/EVP_CIPHER objects before using them
    - in some cases when there is an internal fetch such as in case of composite algorithms (MACs, KDFs), some performance can be regained by preinitializing the MAC or KDF context with all the parameters set and duplicating it

3. Many more locking calls
    - these are necessary due to possibility of loading/unloading providers dynamically, lazy loading of provider algorithm implementations and caching
    - there are some incremental improvements in 3.1 and 3.2 versions mostly by removing the write lock contention, but there are still many more read locks in critical paths than in 1.1.1

4. Decoder (and little bit less importantly encoder) performance issues
    - apart from some optimizations done in 3.1 version and further in 3.2 version the way the decoding of certificates and keys is currently much more complex process than in 1.1.1
    - it is probably possible to optimize this process further but it will require thorough refactoring
    - application workarounds are possible but they basically require to avoid loading the same certificate and private key data repeatedly - i.e., load once, use multiple times

5. OSSL_PARAM handling
     - this is a problem especially in the most performance critical workloads such as symmetric crypto operations - encryptions/decryptions/digests/macs/kdfs
     - some optimizations were done in 3.1
     - handling of OSSL_PARAMs in AEAD encryption was improved in 3.2
     - further optimizations are possible for other types of symmetric algorithms

**Comment by richsalz** [comment]
You should track down the open issues and link them here, ideally with summaries.  (Good luck on that, some of those threads area long)

**Comment by davidben** [comment]
> these are necessary due to possibility of loading/unloading providers dynamically, lazy loading of provider algorithm implementations and caching

I don't think it's right to describe this as *necessary*. First, lazy loading and caching are merely an implementation decision for your underlying design. And indeed, given the performance issues that have come up, I don't think it was the right one. https://github.com/openssl/openssl/issues/20715#issuecomment-1731862606 sketches out a much simpler design which would have limited the thread contention to just the mutable provider list. It would require rewriting a decent amount of the provider internals, but perhaps not a bad thing given the perf issues that have come up.

From there, there's the discussion of provider unload, also in that issue. Provider unload significantly adds to the thread condition that you all face and really should be reconsidered.

Finally, there's the "freeze" idea, which helps with the mutable provider list. I think the way to think about this is that the original design decision to make everything dynamic was perhaps not right. A freeze API is imperfect (callers need to remember to call it), but it is an effective way to meet backwards compatibility while partially mitigating an unfortunate design. It can be a stopgap until 4.0, when you all can reconsider some of the public API decisions that led to this slate of issues.

For example, the use cases behind libctx mutability can also be met by saying callers create new libctxs when they wish to change the set of algorithms available.

I've generally noticed this as a pattern: OpenSSL often makes mutable APIs with object reuse, rather than immutable with object replacement. I filed #5158 back in the 1.1.x days, to reconsider for what became 3.x. At the time, the provider issues had not been introduced, so I was focused on impacts on the `RSA` object, but it's the same thing: mutability and piecemeal initialization in the `RSA` object causes thread contention in shared `RSA` objects, just as mutability and piecemeal initialization in OpenSSL 3.x libctx and providers cause thread contention. This class of 3.x performance problems are ultimately a more extreme version of the same effect, because libctxs are shared across threads even more than `RSA` objects were.

> it is probably possible to optimize this process further but it will require thorough refactoring

https://github.com/openssl/openssl/issues/15199#issuecomment-1631038583 discusses this problem. It's not so much that the current parser is unoptimized. Rather, the current parser does not match how these formats were designed to be parsed in the first place. If you all replace it with the more straightforward design, that problem would entirely go away, and it would still be as pluggable as before.

**Reply by t8m** [reply]
I meant they are necessary given the possibility of loading/unloading and even the no-cache flag on fetched algorithms from providers. Yes, these constraints can be changed but not by 3.x except for the freezing concept.

**Reply by davidben** [reply]
https://github.com/openssl/openssl/issues/20715#issuecomment-1731862606 is compatible with unloading. The no-cache flag is a little interesting and may require some wrangling. When I looked at it at the time, semantics were pretty unclear. Object lifetime is security critical in languages like C (memory safety), yet the docs only say "If it is permissible to cache references to this array then `*no_store` [typo for `*no_cache`?] is set to 0 or 1 otherwise. If the array is not cacheable then it is assumed to have a short lifetime."

Regardless, I expect the more performant design can be made to coexist with no-cache. I.e. I don't believe *all* the thread contention that OpenSSL 3.x introduced was necessary, even under the constraints of the API decisions made in OpenSSL 3.x. A lot of it seems to come from implementation decisions.

But you're certainly right that some of the lock contention _does_ come from API decisions, mostly in the same vein as #5158, just applied to an even more impactful place. I filed #5158 with the hope that you all would keep it in mind in designing 1.1.x's successor. We're now at 1.1.x's successor (3.x), so that didn't happen. Perhaps it can happen for 3.x's successor?
